### PR TITLE
Relax requirements on index signatures to 'any' when a type also contains a string index signature to 'any'

### LINKED ISSUE #43064: Relax 'any' index signature requirements when a type contains a string index signature to 'any'
@DanielRosenwasser Wes and I spoke and recommend that we keep the assignability-target rule in place when the target has both string and numeric index signatures

_Originally posted by @RyanCavanaugh in https://github.com/microsoft/TypeScript/issues/43004#issuecomment-789319883_

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #43064
____

When checking compatibility between two types, TypeScript should not require a corresponding source index signature if the target side index signature maps to `any` *and* the target side has a string index signature that maps to `any`.

So an index signature like in

```
{ [x: number]: any }
```

is still required of a source type, but neither index signature in

```
{ [x: number]: any, [x: string]: any; }
```

should be required; *however*, the number index signature in

```
{ [x: number]: number, [x: string]: any; }
```

should always be required because the number index signature doesn't map to `any`.

_____

TL;DR:

Given `type SomeObject = { hello: string, world: number }`:

* ✅ `SomeObject -> { [x: string]: any }`
* ✅ `SomeObject -> { [x: string]: any, [x: number]: any }`
* ❌ `SomeObject -> { [x: number]: any }`
* ❌ `SomeObject -> { [x: string]: any, [x: number]: string }`

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
One thing I just thought of - when #26797, I think it would be best *not* to relax the requirements to other index signatures since this is just meant to unbreak code.

File: tests/baselines/reference/objectTypeWithStringAndNumberIndexSignatureToAny.errors.txt [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        ~~~~~~
+!!! error TS2322: Type 'Obj' is not assignable to type 'NumberTo<any>'.
+    
+        bothToAny = sToAny;
+        bothToAny = nToAny;
+        bothToAny = someObj;
+        ~~~~~~~~~
+!!! error TS2322: Type 'Obj' is not assignable to type 'StringTo<any> & NumberTo<any>'.
+!!! error TS2322:   Type 'Obj' is not assignable to type 'NumberTo<any>'.
+!!! error TS2322:     Index signature is missing in type 'Obj'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/objectTypeWithStringAndNumberIndexSignatureToAny.errors.txt:101]: This might be questionable. This used to work in 4.2, but this PR doesn't un-break this code.

File: tests/baselines/reference/objectTypeWithStringAndNumberIndexSignatureToAny.errors.txt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        nToAny = sToAny;
+        nToAny = bothToAny;
+        nToAny = someObj;
+        ~~~~~~
+!!! error TS2322: Type 'Obj' is not assignable to type 'NumberTo<any>'.
+!!! error TS2322:   Index signature is missing in type 'Obj'.
+    
+        bothToAny = sToAny;
+        bothToAny = nToAny;
+        bothToAny = someObj;
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/objectTypeWithStringAndNumberIndexSignatureToAny.errors.txt:71]: This is the example that's been un-broken.
